packages/store-sdk/src/cart/Cart.tsx

```diff
@@ -0,0 +1,185 @@
+  price: number
+  listPrice: number
+  quantity: number
+  giftQuantity: number
+}
+
+export interface Cart {
+  id: string
+  items: Record<Item['id'], Item>
+}
+
+export interface ContextValue<T extends Item = Item>
+  extends Omit<Cart, 'items'> {
+  items: T[]
+  gifts: T[]
+  isEmpty: boolean
+  totalItems: number
+  totalUniqueItems: number
+  total: number
+  subTotal: number
+  addItem: (item: T) => void
+  removeItem: (id: Item['id']) => void
+  updateItemQuantity: (id: Item['id'], quantity: number) => void
+  emptyCart: () => void
+  getItem: (id: Item['id']) => T | undefined
+  inCart: (id: Item['id']) => boolean
+  setCart: (cart: Omit<Cart, 'items'> & { items: T[] }) => void
+}
+
+export const Context = createContext<ContextValue | undefined>(undefined)
+Context.displayName = 'StoreCartContext'
+
+const initialState: Cart = {
+  id: '',
+  items: {},
+}
+
+const setItemToCartState = (cart: Cart, id: Item['id'], item: Item) => ({
+  ...cart,
+  items: { ...cart.items, [id]: item },
+})
+
+const removeItemFromCartState = (cart: Cart, id: Item['id']) => {
+  const { [id]: removed, ...items } = cart.items
+
+  return {
+    ...cart,
+    items,
+  }
+}
+
+export interface Props {
+  namespace?: string
+  initialCart?: Cart
+  onItemAdd?: (newItem: Item, oldItem?: Item) => void
+  onItemUpdate?: (newItem: Item, oldItem?: Item) => void
+  onItemRemove?: (item?: Item) => void
+}
+
+export const CartProvider: FC<Props> = ({
+  children,
+  onItemAdd,
+  initialCart,
+  onItemRemove,
+  onItemUpdate,
+  namespace: nspc = 'main',
+}) => {
+  const namespace = `${nspc}::store::cart`
+  const [cart, setCart] = useStorage(namespace, {
+    ...initialState,
+    ...initialCart,
+  })
+
+  const value = useMemo((): ContextValue => {
+    const items = Object.values(cart.items)
+
+    const set: ContextValue['setCart'] = (c) => {
+      const newCart = {
+        ...c,
+        items: c.items.reduce((acc, curr) => {
+          acc[curr.id] = curr
+
+          return acc
+        }, {} as Record<string, Item>),
+      }
+
+      setCart(newCart)
+    }
+
+    const addItem = (item: Item) => {
+      if (!item.id) {
+        throw new SDKError('You must provide an `id` for items')
+      }
+
+      if (item.quantity < 0) {
+        throw new SDKError('Item quantity needs to be higher than zero')
+      }
+
+      const currentItem = cart.items[item.id]
+
+      const newItem = currentItem
+        ? {
+            ...item,
+            quantity: currentItem.quantity + item.quantity,
+            giftQuantity: currentItem.giftQuantity + item.giftQuantity,
+          }
+        : item
+
+      const callback = currentItem ? onItemUpdate : onItemAdd
+
+      setCart(setItemToCartState(cart, newItem.id, newItem))
+      callback?.(newItem, currentItem)
+    }
+
+    const updateItemQuantity = (id: Item['id'], quantity: number) => {
+      if (quantity < 0) {
+        const item = cart.items[id]
+
+        setCart(removeItemFromCartState(cart, id))
```

Comment:
This would remove the gifts as well, right? What makes it impossible for users to update the gift quantity unless they add a new gift item. But still, they wouldn't be able to decrease it unless they sent a negative value as the item's `giftQuantity` to the `addItems` function, which sounds wrong.

Also, have you considered an update item function? One that doesn't only change the quantity, but any property of the product? I think this might come in handy for some cases.

Reply:
Yes, I have removed this section. If you want to remove an item or change another property, you need to remove then add it again